src/rest/connector/acisdk_implementation.py

```diff
@@ -0,0 +1,285 @@
+from functools import wraps
+from importlib import import_module
+from logging import getLogger
+from pprint import pformat
+from re import search
+import urllib3
+
+from pyats.connections import BaseConnection
+from rest.connector.utils import get_username_password, verify_apic_version
+
+# create a logger for this module
+log = getLogger(__name__)
+
+
+class AciCobra(BaseConnection):
+    """ACI SDK (Cobra) Implementation for APIC
+
+    Implementation of ACI SDK connection to devices
+    based on pyATS BaseConnection for APIC
+
+    YAML Example
+    ------------
+
+        devices:
+            apic1:
+                connections:
+                    cobra:
+                        class: rest.connector.Acisdk
+                        ip : "2.3.4.5"
+                credentials:
+                    cobra:
+                        username: admin
+                        password: cisco123
+
+    Code Example
+    ------------
+
+        >>> from pyats.topology import loader
+        >>> testbed = loader.load('/users/xxx/xxx/testbed.yaml')
+        >>> device = testbed.devices['apic1']
+        >>> device.connect(alias='cobra', via='cobra')
+        >>> device.rest.connected
+        True
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.mo_dir = None
+
+        # remove warnings for insecure HTTPS
+        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+
+        self._is_connected = False
+
+        if 'host' in self.connection_info:
+            ip = self.connection_info['host']
+        else:
+            ip = self.connection_info['ip'].exploded
+        if 'port' in self.connection_info:
+            port = self.connection_info['port']
+            self.url = f'https://{ip}:{port}/'
+        else:
+            self.url = f'https://{ip}/'
+
+        verify_apic_version(ip)
+
+    @property
+    def connected(self):
+        """Is a device connected"""
+        return self._is_connected
+
+    @BaseConnection.locked
+    def connect(self, timeout=90):
```

Comment:
Please add Unittest for this connector

Reply:
waiting for the update. here is example of UT from APIC connector. https://github.com/CiscoTestAutomation/rest/blob/master/src/rest/connector/tests/test_apic.py

Reply:
I have been thinking about this, and I have to admit I'm at a loss.
the connector relies heavily on a set of libraries obtained from the APIC itself - while, in theory, it's possible to craft the responses normally given by the device, I fear the unit test will fail when the libraries won't be present.
Honestly, I currently don't know how to circumvent this problem

Reply:
ok. please write UT with Cobra libraries. and then add logic in UT to skip the tests if the library is not installed.

Reply:
Test added under src/rest/connector/tests/test_apic_cobra.py

Reply:
thanks